README.md
# ember-angle-bracket-invocation-polyfill

This addon provides a polyfill for angle bracket invocation syntax as described
in [RFC 311](https://github.com/emberjs/rfcs/pull/311) and [RFC
457](https://emberjs.github.io/rfcs/0457-nested-lookups.html). It's the same
components you know and love, no longer surrounded by mustaches. \o/

[![Build Status](https://travis-ci.org/rwjblue/ember-angle-bracket-invocation-polyfill.svg?branch=master)](https://travis-ci.org/rwjblue/ember-angle-bracket-invocation-polyfill)

## Installation

```
ember install ember-angle-bracket-invocation-polyfill
```

You will additionally need to ensure ember-cli-htmlbars-inline-precompile is at least version 1.0.3.

## Usage

The best usage guide are the RFCs themselves
([emberjs/rfcs#311](https://emberjs.github.io/rfcs/0311-angle-bracket-invocation.html)
[emberjs/rfcs#457](https://emberjs.github.io/rfcs/0457-nested-lookups.html)),
but here are a few examples of "before"/"after" to whet your appetite:

**Before**:

```hbs
{{site-header user=this.user class=(if this.user.isAdmin "admin")}}

{{#super-select selected=this.user.country as |s|}}
  {{#each this.availableCountries as |country|}}
    {{#s.option value=country}}{{country.name}}{{/s.option}}
  {{/each}}
{{/super-select}}
```

**After**:

```hbs
<SiteHeader @user={{this.user}} class={{if this.user.isAdmin "admin"}} />

<SuperSelect @selected={{this.user.country}} as |s|>
  {{#each this.availableCountries as |country|}}
    <s.option @value={{country}}>{{country.name}}</s.option>
  {{/each}}
</SuperSelect>
```

### Supported Features

- Supports invoking components via angle brackets using TitleCase

```hbs
<FooBar></FooBar>
```

- Supports invoking components via angle brackets with self-closing syntax

```hbs
<FooBar />
```

- Supports invoking components via angle brackets using paths

```hbs
<some.thing></some.thing>
```

- Supports invoking components via angle brackets using yielded block params

```hbs
{{#with (component 'foo-bar') as |Foo|}}
  <Foo></Foo>
{{/with}}
```

- Supports passing arguments into invoked components with `@` prefix

```hbs
<FooBar @title={{whateverHere}}></FooBar>
```

- Supports passing attributes to be added to the root element of the component without `@` prefix

```hbs
<FooBar data-test-foo-bar></FooBar>
```

- Supports using `has-block` _inside_ the invoked component to determine if a block was provided

```hbs
<FooBar /> {{! checking `has-block` inside would be `false`}}

<FooBar></FooBar> {{! checking `has-block` inside would be `true`}}
```

- Supports yielding block params from inside the invoked component

```hbs
<FooBar as |qux|>{{qux}}</FooBar>
```

- Supports passing `...attributes` into another angle bracket invocation

```hbs
<FooBar ...attributes>
```

- Supports passing `...attributes` into a non-component element

```hbs
<div ...attributes></div>
```

- Supports passing `...attributes` into tagless components
- Supports invoking single word components:

```hbs
<Title />
```

- Supports invoking components nested in subfolders:

```
<Foo::Bar />
```

- Completely inert when running Ember 3.10 or higher
- Supports Ember 2.12, 2.16, 2.18, 3.1, 3.2, 3.3, 3.4, 3.8, 3.9
- Test all the features listed above 😘

## Limitations

There are no known limitations, all features described in the RFC are polyfilled.

## Contributing

### Installation

- `git clone <repository-url>`
- `cd ember-angle-bracket-invocation-polyfill`
- `yarn install`

### Linting

- `yarn lint:js`
- `yarn lint:js --fix`

### Running tests

- `ember test` – Runs the test suite on the current Ember version
- `ember test --server` – Runs the test suite in "watch mode"
- `ember try:each` – Runs the test suite against multiple Ember versions

### Running the dummy application

- `ember serve`
- Visit the dummy application at [http://localhost:4200](http://localhost:4200).

For more information on using ember-cli, visit [https://ember-cli.com/](https://ember-cli.com/).

## License

This project is licensed under the [MIT License](LICENSE.md).


CHANGELOG.md
## v3.0.2 (2022-05-18)

#### :bug: Bug Fix
* [#128](https://github.com/ember-polyfills/ember-angle-bracket-invocation-polyfill/pull/128) Get CI passing ([@mmun](https://github.com/mmun))
* [#126](https://github.com/ember-polyfills/ember-angle-bracket-invocation-polyfill/pull/126) Fix handling of attributes in Ember 3.4 ([@twokul](https://github.com/twokul))

#### Committers: 1
- Martin Muñoz ([@mmun](https://github.com/mmun))
- Alex Navasardyan ([@twokul](https://github.com/twokul))


## v3.0.1 (2021-04-23)

#### :bug: Bug Fix
* [#125](https://github.com/ember-polyfills/ember-angle-bracket-invocation-polyfill/pull/125) Fix monkey patching in Ember 2.x polyfill code paths ([@mmun](https://github.com/mmun))

#### Committers: 1
- Martin Muñoz ([@mmun](https://github.com/mmun))


## v3.0.0 (2020-10-09)

#### :boom: Breaking Change
* [#124](https://github.com/ember-polyfills/ember-angle-bracket-invocation-polyfill/pull/124) Drop Node < 10 support; Upgrade ember-cli-babel to v7 ([@nlfurniss](https://github.com/nlfurniss))

#### Committers: 1
- Nathaniel Furniss ([@nlfurniss](https://github.com/nlfurniss))


## v2.1.0 (2020-06-05)

#### :rocket: Enhancement
* [#91](https://github.com/ember-polyfills/ember-angle-bracket-invocation-polyfill/pull/91) Ensure polyfill works for Engines ([@pgengler](https://github.com/pgengler))

#### :memo: Documentation
* [#90](https://github.com/ember-polyfills/ember-angle-bracket-invocation-polyfill/pull/90) Update README with a section on usage within addons ([@Techn1x](https://github.com/Techn1x))

#### Committers: 4
- Brad Overton ([@Techn1x](https://github.com/Techn1x))
- David Tang ([@skaterdav85](https://github.com/skaterdav85))
- Phil Gengler ([@pgengler](https://github.com/pgengler))
- Robert Jackson ([@rwjblue](https://github.com/rwjblue))

## v2.0.2 (2019-08-06)

#### :bug: Bug Fix
* [#82](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/82) Fix <LinkTo> usage with `class` attribute. ([@Alonski](https://github.com/Alonski))

#### :house: Internal
* [#84](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/84) Refresh `yarn.lock` ([@simonihmig](https://github.com/simonihmig))

#### Committers: 2
- Alon Bukai ([@Alonski](https://github.com/Alonski))
- Simon Ihmig ([@simonihmig](https://github.com/simonihmig))

## v2.0.1 (2019-08-02)

#### :bug: Bug Fix
* [#79](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/79) Fix transform to not touch native <input> ([@simonihmig](https://github.com/simonihmig))

#### Committers: 1
- Simon Ihmig ([@simonihmig](https://github.com/simonihmig))

## v2.0.0 (2019-08-02)

#### :boom: Breaking Change
* [#74](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/74) Drop node 6, 9, and 11 support. ([@simonihmig](https://github.com/simonihmig))

#### :rocket: Enhancement
* [#75](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/75) Add polyfill for <Input> and <Textarea> ([@simonihmig](https://github.com/simonihmig))
* [#72](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/72) Add polyfill for angle brackets <LinkTo> ([@simonihmig](https://github.com/simonihmig))

#### :bug: Bug Fix
* [#77](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/77) Ensure unsupported arguments to `<Input>` and `<LinkTo>` issue warnings or errors appropriately ([@simonihmig](https://github.com/simonihmig))
* [#76](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/76) Add assertion for not passing @model and @models to <LinkTo> ([@simonihmig](https://github.com/simonihmig))

#### :house: Internal
* [#73](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/73) Drop ember-cli-eslint in favor of plain eslint ([@simonihmig](https://github.com/simonihmig))
* [#63](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/63) Run class merging test on 3.9.1. ([@rwjblue](https://github.com/rwjblue))

#### Committers: 2
- Robert Jackson ([@rwjblue](https://github.com/rwjblue))
- Simon Ihmig ([@simonihmig](https://github.com/simonihmig))

## v1.3.1 (2019-05-08)

#### :bug: Bug Fix
* [#68](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/68) Fix error when using splattributes but no attributes were passed ([@simonihmig](https://github.com/simonihmig))

#### Committers: 1
- Simon Ihmig ([@simonihmig](https://github.com/simonihmig))

## v1.3.0 (2019-04-08)

#### :rocket: Enhancement
* [#59](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/59) Add support for nested invocation with `::` ([@rtablada](https://github.com/rtablada))

#### :house: Internal
* [#60](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/60) Add ember-lts-3.8 to CI setup. ([@rwjblue](https://github.com/rwjblue))

#### Committers: 2
- Robert Jackson ([@rwjblue](https://github.com/rwjblue))
- Ryan Tablada ([@rtablada](https://github.com/rtablada))

## v1.2.5 (2019-01-31)

#### :bug: Bug Fix
* [#56](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/56) Fix class merging and attribute precedence ([@ef4](https://github.com/ef4))

#### Committers: 1
- Edward Faulkner ([@ef4](https://github.com/ef4))


## v1.2.4 (2019-01-27)

#### :bug: Bug Fix
* [#55](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/55) Fix for max call stack size exceeded when runing many tests ([@thousand](https://github.com/thousand))

#### :memo: Documentation
* [#46](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/46) Update Changelog ([@Turbo87](https://github.com/Turbo87))

#### :house: Internal
* [#54](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill/pull/54) TravisCI: Remove deprecated `sudo: false` option ([@Turbo87](https://github.com/Turbo87))

#### Committers: 2
- Bryan Levay ([@thousand](https://github.com/thousand))

config/ember-try.js
'use strict';

const getChannelURL = require('ember-source-channel-url');

module.exports = function() {
  return Promise.all([
    getChannelURL('release'),
    getChannelURL('beta'),
    getChannelURL('canary'),
  ]).then(() => {
    return {
      useYarn: true,
      scenarios: [
        {
          name: 'ember-lts-2.12',
          npm: {
            devDependencies: {
              'ember-source': '~2.12.0',
            },
          },
        },
        {
          name: 'ember-lts-2.16',
          npm: {
            devDependencies: {
              'ember-source': '~2.16.0',
            },
          },
        },
        {
          name: 'ember-lts-2.18',
          npm: {
            devDependencies: {
              'ember-source': '~2.18.0',
            },
          },
        },
        {
          name: 'ember-3.1',
          npm: {
            devDependencies: {
              'ember-source': '~3.1.0',
            },
          },
        },
        {
          name: 'ember-3.2',
          npm: {
            devDependencies: {
              'ember-source': '~3.2.0',
            },
          },
        },
        {
          name: 'ember-3.3',
          npm: {
            devDependencies: {
              'ember-source': '~3.3.0',
            },
          },
        },
        {
          name: 'ember-lts-3.4',
          npm: {
            devDependencies: {
              'ember-source': '~3.4.0',
            },
          },
        },
        {
          name: 'ember-lts-3.8',
          npm: {
            devDependencies: {
              'ember-source': '~3.8.0',
            },
          },
        },
        {
          name: 'ember-lts-3.28',
          npm: {
            devDependencies: {
              'ember-source': '~3.28.0',
            },
          },
        },
        {
          name: 'ember-default',
          npm: {
            devDependencies: {},
          },
        },
      ],
    };
  });
};
